### PR TITLE
Add a CLI `iree-ir-tool` with a command to strip data.

### PR DESCRIPTION
**Pull request by stellaraccident**
Been meaning to do this for a while in order to have a place to stash more power-user style things that core developers typically use iree-opt for.

This one adds a `strip-data` sub-command which uses the passes from #14627 to systematically replace tensor constants with synthetic values. With an ASAN/asserts build, this was able to strip a 7GiB int4 vicuna MLIR file in ~5s and a 23GiB h2ogpt model in about 30s (the latter has some other characteristics which make it more expensive to load as well as being bigger). Results were a 2.6MiB and 1.4MiB MLIR file respectively, consisting just of program IR and annotations for synthetic data.

Getting the opt pipeline right for arbitrary input is a bit tricky, so I decided we should just armor this into a tool

From installed packages, this can be used as:

```
iree-ir-tool strip-data input.mlir -o output.mlir
```

From a build tree with Python setup:

```
python -m iree.compiler.tools.ir_tool strip-data input.mlir -o output.mlir
```

Required adding some additional compiler APIs:

* `ireeCompilerInvocationRunPassPipeline` to run an arbitrary textual pass pipeline on an invocation.
* `ireeCompilerInvocationOutputIRBytecode` to emit bytecode from an invocation.

### CODE REVIEW

**Review by benvanik — APPROVED** [review]
beautiful!

File: compiler/bindings/python/test/tools/ir_tool_test.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from iree.compiler.tools.ir_tool import __main__
```
Comment (ScottTodd) [inline compiler/bindings/python/test/tools/ir_tool_test.py:7]: This test failed on the Windows CI:
https://github.com/openxla/iree/actions/runs/5829685985/job/15809623394#step:9:2055
```
 962/1162 Test #1331: iree/compiler/bindings/python/test/tools/ir_tool_test ......................................................***Failed    1.12 sec
Traceback (most recent call last):
  File "C:\a\iree\iree\compiler\bindings\python\test\tools\ir_tool_test.py", line 7, in <module>
    from iree.compiler.tools.ir_tool import __main__
  File "C:\a\iree\iree\build-windows\compiler\bindings\python\iree\compiler\tools\ir_tool\__main__.py", line 11, in <module>
    from ...api import Invocation, Session, Source, Output
  File "C:\a\iree\iree\build-windows\compiler\bindings\python\iree\compiler\api\__init__.py", line 26, in <module>
    from .ctypes_dl import (
  File "C:\a\iree\iree\build-windows\compiler\bindings\python\iree\compiler\api\ctypes_dl.py", line 454, in <module>
    _global_init = _GlobalInit()
  File "C:\a\iree\iree\build-windows\compiler\bindings\python\iree\compiler\api\ctypes_dl.py", line 443, in __init__
    _init_dylib()
  File "C:\a\iree\iree\build-windows\compiler\bindings\python\iree\compiler\api\ctypes_dl.py", line 40, in _init_dylib
    dylib_path = _probe_iree_compiler_dylib()
  File "C:\a\iree\iree\build-windows\compiler\bindings\python\iree\compiler\api\ctypes_dl.py", line 438, in _probe_iree_compiler_dylib
    raise ValueError(f"Could not find {dylib_basename} in {paths}")
ValueError: Could not find IREECompiler.dll in [WindowsPath('C:/a/iree/iree/build-windows/lib')]
Exception ignored in: <function _GlobalInit.__del__ at 0x000001A5E8B33520>
Traceback (most recent call last):
  File "C:\a\iree\iree\build-windows\compiler\bindings\python\iree\compiler\api\ctypes_dl.py", line 450, in __del__
    self.local_dylib.ireeCompilerGlobalShutdown()
AttributeError: '_GlobalInit' object has no attribute 'local_dylib'
```

Comment (stellaraccident) [reply compiler/bindings/python/test/tools/ir_tool_test.py:7]: Oh, I think I had a patch for that on a branch somewhere but never landed. Will fix.

Comment (stellaraccident) [reply compiler/bindings/python/test/tools/ir_tool_test.py:7]: Direct pushed a fix: https://github.com/openxla/iree/commit/8ab9e12b0bf7721a648710efbdb24db41f962c10

Comment (ScottTodd) [reply compiler/bindings/python/test/tools/ir_tool_test.py:7]: Thanks!

**Pull request merged**